tests/Unit/DataStructures/Tensor/Expressions/Test_Evaluate.cpp

```diff
@@ -0,0 +1,268 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "Helpers/DataStructures/Tensor/Expressions/EvaluateRank0TestHelpers.hpp"
+#include "Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp"
+#include "Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp"
+#include "Helpers/DataStructures/Tensor/Expressions/EvaluateRank3TestHelpers.hpp"
+#include "Helpers/DataStructures/Tensor/Expressions/EvaluateRank4TestHelpers.hpp"
+
+namespace {
+template <auto&... TensorIndices>
+void test_contains_indices_to_contract_impl(const bool expected) {
+  CHECK(TensorExpressions::detail::contains_indices_to_contract<sizeof...(
+            TensorIndices)>(
+            {{std::decay_t<decltype(TensorIndices)>::value...}}) == expected);
+}
+
+void test_contains_indices_to_contract() {
+  test_contains_indices_to_contract_impl<ti_a, ti_b, ti_c>(false);
+  test_contains_indices_to_contract_impl<ti_I, ti_j>(false);
+  test_contains_indices_to_contract_impl<ti_j>(false);
+  test_contains_indices_to_contract_impl(false);
+
+  test_contains_indices_to_contract_impl<ti_d, ti_D>(true);
+  test_contains_indices_to_contract_impl<ti_I, ti_i>(true);
+  test_contains_indices_to_contract_impl<ti_a, ti_K, ti_B, ti_b>(true);
+  test_contains_indices_to_contract_impl<ti_j, ti_c, ti_J, ti_A, ti_a>(true);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Expression.Evaluate",
+                  "[DataStructures][Unit]") {
+  INFO("Testing contains_indices_to_contract");
```

Comment:
Just an FYI, `INFO`s are scoped so if you put this _inside_ the `test_contains...` function it would only get printed if one of the checks inside that function (including nested calls) fail :)

Reply:
Oh I see! So then should I just remove it altogether? I'm not sure it's adding anything informative to the output for this test when a test case fails:
```
/home/alexmacedo/spectre/tests/Unit/DataStructures/Tensor/Expressions/Test_Evaluate.cpp:20: FAILED:
  CHECK( TensorExpressions::detail::contains_indices_to_contract<sizeof...( TensorIndices)>( {{std::decay_t<decltype(TensorIndices)>::value...}}) == expected )
with expansion:
  false == true
with message:
  Testing contains_indices_to_contract
```
^ I just purposely passed in the wrong `expected` value in a call to see what the output would look like

Reply:
What I'm saying is that currently all `CHECK`s that fail even after line 39 will have this message printed and that could be misleading since it's not longer checking `contains_indices_to_contract`. However, I _do_ think it's good to use `INFO`s to give extra context. If you put the `INFO` inside the `test_contains_indices_to_contract` function, then the message will not be printed on checks below line 39. Does that make sense? And again, this is just an FYI, don't feel obligated to change code if you don't see it as a useful thing to change :)

Reply:
Oh, sorry!  Yes, this all made/makes sense to me - if a `test_evaluate_rank_#` `CHECK` failed, the output would be confusing. I think the `INFO` should not be floating out in the `SPECTRE_TEST_CASE`. So my comment/question above was intended to be: yes, I see now why it should be inside the function instead of the `SPECTRE_TEST_CASE`, but now I am wondering if it is adding anything useful to the output, at all, or if I should remove it. In the output above, it doesn't seem to me like it's providing any extra information, and I am in favor of just not using it at all, given that the test case and function are both small and straight-forward. On the contrary, I could see `INFO` being super useful for a more complex test or function, where I'm logging multiple things to help debug.

Let me know your thoughts!

Reply:
So I think it definitely needs to be removed from the scope of the `SPECTRE_TEST_CASE` based on the scoping you mentioned, but now that I know it's only printed when a `CHECK` fails, I'm considering just not using `INFO` for this at all since I'm not sure it's adding much to the failure output. Sorry I didn't explain that clearer the first time!

Reply:
Yea, it's not clear how helpful it is here. It's sometimes good if there is a helper function that gets called in a few very different places, which means the `CHECK`s inside will report a not so helpful line number (you don't know where the wrapping function was called from), and in that case using `INFO` to add some additional output is useful. So here, I'm fine with leaving it, removing it, or scoping it, whichever you prefer :P

Reply:
Ok! I've gone ahead and removed it, but I'm really glad to know about `INFO` so that I can utilize it in other tests. Thanks!